connector/grafanacloudconnector/connector.go

```diff
@@ -0,0 +1,160 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package grafanacloudconnector // import "github.com/open-telemetry/opentelemetry-collector-contrib/connector/grafanacloudconnector"
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/connector"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	"go.opentelemetry.io/otel/metric"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/connector/grafanacloudconnector/internal/metadata"
+)
+
+const (
+	hostInfoMetric     = "traces_host_info"
+	hostIdentifierAttr = "grafana.host.id"
+)
+
+var _ connector.Traces = (*connectorImp)(nil)
+
+type connectorImp struct {
+	config Config
+	logger *zap.Logger
+
+	started      bool
+	done         chan struct{}
+	shutdownOnce sync.Once
+
+	metricsConsumer consumer.Metrics
+	hostMetrics     *hostMetrics
+
+	metricHostCount      metric.Int64UpDownCounter
+	metricFlushCount     metric.Int64Counter
+	metricDatapointCount metric.Int64Counter
+}
+
+func newConnector(logger *zap.Logger, set component.TelemetrySettings, config component.Config) (*connectorImp, error) {
+	mHostCount, err := metadata.Meter(set).Int64UpDownCounter(
+		"grafanacloud_host_count",
+		metric.WithDescription("Number of unique hosts since last metrics flush"),
+		metric.WithUnit("1"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	mFlushCount, err := metadata.Meter(set).Int64Counter(
+		"grafanacloud_flush_count",
+		metric.WithDescription("Number of metrics flushes"),
+		metric.WithUnit("1"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	mDatapointCount, err := metadata.Meter(set).Int64Counter(
+		"grafanacloud_datapoint_count",
+		metric.WithDescription("Number of datapoints sent to Grafana Cloud"),
+		metric.WithUnit("1"),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	cfg := config.(*Config)
+	return &connectorImp{
+		config:               *cfg,
+		logger:               logger,
+		done:                 make(chan struct{}),
+		hostMetrics:          newHostMetrics(),
+		metricHostCount:      mHostCount,
+		metricFlushCount:     mFlushCount,
+		metricDatapointCount: mDatapointCount,
+	}, nil
+}
+
+// Capabilities implements connector.Traces.
+func (c *connectorImp) Capabilities() consumer.Capabilities {
+	return consumer.Capabilities{MutatesData: false}
+}
+
+// ConsumeTraces implements connector.Traces.
+func (c *connectorImp) ConsumeTraces(_ context.Context, td ptrace.Traces) error {
+	for i := 0; i < td.ResourceSpans().Len(); i++ {
+		resourceSpan := td.ResourceSpans().At(i)
+		attrs := resourceSpan.Resource().Attributes()
+		mapping := attrs.AsRaw()
+
+		for _, attrName := range c.config.HostIdentifiers {
+			if val, ok := mapping[attrName]; ok {
+				if v, ok := val.(string); ok {
+					c.hostMetrics.add(v)
+					c.metricHostCount.Add(context.Background(), int64(1))
```

Comment:
I put some more thought into this metric and this isn't the right spot to count these hosts. It makes more sense to set the host count when we flush.